src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -717,12 +729,22 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
     expectCompletedMerge ln (SingleRun r) = do
       traceWith tr $ AtLevel ln $ TraceExpectCompletedMergeSingleRun (runNumber $ Run.runRunFsPaths r)
       pure r
-    expectCompletedMerge ln (MergingRun _ _ var) = do
-      r <- withMVar var $ \case
+    expectCompletedMerge ln (MergingRun _ _ unspentCreditsVar mergeVar) = do
+      -- If there are remaining unspent credits that are below the threshold,
+      -- then we might not have stepped the merge to completion yet. We take the
+      -- remaining unspent credits and we /try/ to finish the merge.
+      isMergeDone <-
+        bracketOnError
+          (takeAllUnspentCredits unspentCreditsVar)
+          (putBackUnspentCredits unspentCreditsVar)
+          (stepMerge mergeVar)
+      when isMergeDone $ completeMerge mergeVar
```

Comment:
(knowing how many total steps a merge requires would also help with making `scaleCreditsForMerge` distribute the work more evenly)

Reply:
Regardless of the question whether `expectCompletedMerge` should do the last bit of merging work, I agree that keeping track of the input sizes would make it possible to better distribute merge work over time

I agree that it would be nice to have the prototype and implementation be in sync, though I'm going to play devil's advocate here: is it not the prototype that should change? If we care about dividing batches of work evenly over time, then it actually might make more sense to have `expectCompletedMerge` do the last bit of work (which is below the threshold) instead of having the last update do more work (above the threshold)

Reply:
> having the last update do more work (above the threshold)

The last update wouldn't do more work! The proposed case would trigger when we haven't reached the threshold yet, but it's already enough to complete the merge. So it's less merging work than a usual batch.

Reply:
Doing it in `expectCompletedMerge` even might mean this happens on every level (for a cascade of levels getting full), each potentially doing a little less than a batch, which together could also lead to a little spike.

Reply:
Done